### PR TITLE
new features

### PR DESCRIPTION
**Pull request by ketoo**
add draggable feature for node
add interface to set the node's origin position

### CODE REVIEW

**Comment by Nelarius** [comment]
Thanks for the contribution! 👍  Some thoughts:

* `EditorContextMoveToNode` is a nice idea! Haven't yet personally had node graphs grow that large, but this probably helps out in that case 🙂 
* `SetNodeOriginPos` made me realize that `SetNodePos` isn't clear about what the position is! I think this change is a good addition, but we could name things better. Maybe `SetNodeGridSpacePos` and `SetNodeScreenSpacePos`?
* `SetNodeDraggable` -- it looks like you want to lock certain nodes in place? What sort of use-case do you have in mind?

**Comment by ketoo** [comment]
> Thanks for the contribution! 👍 Some thoughts:
> 
> * `EditorContextMoveToNode` is a nice idea! Haven't yet personally had node graphs grow that large, but this probably helps out in that case 🙂
> * `SetNodeOriginPos` made me realize that `SetNodePos` isn't clear about what the position is! I think this change is a good addition, but we could name things better. Maybe `SetNodeGridSpacePos` and `SetNodeScreenSpacePos`?
> * `SetNodeDraggable` -- it looks like you want to lock certain nodes in place? What sort of use-case do you have in mind?

Thanks for your feedback.

For EditorContextMoveToNode, it's about the user experience, just to do a quick way to let the user can focus on which node now editing.

For SetNodeOriginPos, I will modify the name under your advice and submit a new PR. Is that should be SetNodeScreenSpacePos?

For SetNodeDraggable, we will use it in the case as below: sometimes, we just want to display some items with the node system. For instance, we show some objects which come from other system but the viewer dont have the permission to modify the position as the position come from other systems too.

**Comment by Nelarius** [comment]
> For SetNodeOriginPos, I will modify the name under your advice and submit a new PR. Is that should be SetNodeScreenSpacePos?

Your `SetNodeOriginPos` would be named `SetNodeGridSpacePos` since the node origin is implicitly assumed to be relative to the grid/canvas origin and your function sets this value without any transformations 🙂 It's probably something that I should document in the code a bit better. You can update this PR with the updated name if you like!

I can update the name of existing function 👍 

> For SetNodeDraggable, we will use it in the case as below: sometimes, we just want to display some items with the node system. For instance, we show some objects which come from other system but the viewer dont have the permission to modify the position as the position come from other systems too.

Ok, sounds like a useful feature then!

File: imnodes.cpp [hunk 1]
```diff
@@ -1126,6 +1132,17 @@ void EditorContextResetPanning(const ImVec2& pos)
...
 
+void EditorContextMoveToNode(const int node_id)
+{
+    EditorContext& editor = editor_context_get();
+    NodeData& node = editor.nodes.find_or_create_new(node_id);
+
+    const ImVec2 canvas_size = ImGui::GetWindowSize();
+
+    editor.panning.x = canvas_size.y / 2 - node.origin.x;
+    editor.panning.y = canvas_size.x / 2 - node.origin.y;
```
Comment (Nelarius) [inline imnodes.cpp]: I think `canvas_size.x` and `canvas_size.y` are flipped

File: imnodes.h [hunk 2]
```diff
@@ -127,9 +128,11 @@ void Link(int id, int start_attr, int end_attr);
 // Set's the node's position corresponding to the node id. You can even set the
 // position before the node has been created with BeginNode().
 void SetNodePos(int node_id, const ImVec2& pos);
+void SetNodeOriginPos(int node_id, const ImVec2& origin_pos);
```
Comment (Nelarius) [inline imnodes.h]: Like I already mentioned, this should be named `SetNodeGridSpacePos` 🙂

Comment (ketoo) [reply imnodes.h]: Cool, I will update the PR.

Comment (ketoo) [reply imnodes.h]: Done

**Comment by Nelarius** [comment]
Merged it, and made a few formatting shanges of my own 🙂

**Comment by ketoo** [comment]
> Merged it, and made a few formatting shanges of my own 🙂

Thanks.

**Pull request merged**